gpio/src/gpio.rs

```diff
@@ -0,0 +1,985 @@
+            VIRTIO_GPIO_IRQ_TYPE_EDGE_BOTH => config.set_edge_detection(Edge::Both),
+
+            // Drop the buffer.
+            VIRTIO_GPIO_IRQ_TYPE_NONE => {
+                state.buffer = None;
+                return Ok(());
+            }
+
+            // Only edge IRQs are supported for now.
+            _ => return Err(Error::GpioIrqTypeNotSupported(value)),
+        };
+
+        // Allocate the buffer and configure the line for interrupt.
+        if state.request.is_none() {
+            Err(Error::GpioLineRequestNotConfigured)
+        } else {
+            // The GPIO Virtio specification allows a single interrupt event for each
+            // `wait_for_interrupt()` message. And for that we need a single `EdgeEventBuffer`.
+            state.buffer = Some(EdgeEventBuffer::new(1).map_err(Error::GpiodFailed)?);
+
+            state
+                .request
+                .as_ref()
+                .unwrap()
+                .reconfigure_lines(&config)
+                .map_err(Error::GpiodFailed)
+        }
+    }
+
+    fn wait_for_interrupt(&self, gpio: u16) -> Result<()> {
+        // While waiting here for the interrupt to occur, it is possible that we receive another
+        // request from the guest to disable the interrupt instead, via a call to `set_irq_type()`.
+        //
+        // The interrupt design here should allow that call to return as soon as possible, after
+        // disabling the interrupt and at the same time we need to make sure that we don't end up
+        // freeing resources currently used by `wait_for_interrupt()`.
+        //
+        // To allow that, the line state management is done via two resources: `request` and
+        // `buffer`.
+        //
+        // The `request` is required by `wait_for_interrupt()` to query libgpiod and must not get
+        // freed while we are waiting for an interrupt. This can happen, for example, if another
+        // thread disables the interrupt, via `set_irq_type(VIRTIO_GPIO_IRQ_TYPE_NONE)`, followed
+        // by `set_direction(VIRTIO_GPIO_DIRECTION_NONE)`, where we drop the `request`. For this
+        // reason, the `request` is implemented as an Arc instance.
+        //
+        // The `buffer` on the other hand is required only after we have sensed an interrupt and
+        // need to read it. The design here takes advantage of that and allows `set_irq_type()` to
+        // go and free the `buffer`, while this routine is waiting for the interrupt. Once the
+        // waiting period is over or an interrupt is sensed, `wait_for_interrupt() will find the
+        // buffer being dropped and return an error which will be handled by
+        // `Controller::wait_for_interrupt()`.
+        //
+        // This design also allows `wait_for_interrupt()` to not take a lock for the entire
+        // duration, which can potentially also starve the other thread trying to disable the
+        // interrupt.
+        let request = {
```

Comment:
This looks really good now - all the pieces of the puzzled are clearly explained.

Reply:
Thanks.